kubelet: check that the pod is finished before removing the container.

AnishShah opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When pods with ephemeral storage limits exceed their capacity, the kubelet tries to evict and kill the pod gracefully. We noticed that the pod status changes to `ContainerStatusUnknown`. This is due to a race condition removing the container before pod status can be updated.

In this change, we try to ensure that the pod is finished before removing the container.


#### Which issue(s) this PR fixes:

Fixes #122160 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Fixes an issue where the pod status changes to ContainerStatusUnknown when it is evicted due to exceeding the ephemeral storage limit.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


```docs

```

kannon92 commented:
/sig node

AnishShah commented:
/retest

AnishShah commented:
>  Would this PR need a changelog entry? I think the behavior of the cluster would change.

@sftim, my understanding is that we are not seeing the expected behavior due to a race condition. So it is not changing the behavior but fixing the issue.

sftim commented:
If we fix a bug, we usually add a change log entry. (If there is nothing to fix, does it need merging?)

yujuhong commented:
@SergeyKanzhelev and @bobbypage to help review the change

ndixita commented:
cc: @ndixita

ndixita commented:
/assign @SergeyKanzhelev 
/assign @bobbypage 
/assign @mrunalp

AnishShah committed:
kubelet: check that the pod is finished before removing the container.

When pods with ephemeral storage limits exceed their capacity, the
kubelet tries to evict and kill the pod gracefully. We noticed that the
pod status changes to ContainerStatusUnknown. This is due to a race
condition removing the container before pod status can be updated.

In this change, we try to ensure that the pod is finished before
removing the container.

AnishShah commented:
/retest

AnishShah commented:
I modified the existing eviction test to check the terminated reason is not `ContainerStatusUnknown`. I ran the test multiple times with this command `ginkgo ./test/e2e -focus=".*evicted pods should be terminal.*" -repeat=100` in clusters with and without this fix. The test failed in the cluster without the fix after 22nd attempts whereas the test did not fail in the cluster with this fix.

smarterclayton commented on test/e2e/node/pods.go:
```diff
@@ -339,7 +339,10 @@ var _ = SIGDescribe("Pods Extended", func() {
 				framework.Failf("error waiting for pod to be evicted: %v", err)
 			}
-
+			err = e2epod.EnsureValidContainerTerminatedReason(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "bar", 2*time.Minute)
```
This is an excellent e2e test, and I agree we should have this here.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -690,7 +690,7 @@ func (p *podWorkers) ShouldPodContentBeRemoved(uid types.UID) bool {
 	defer p.podLock.Unlock()
 	if status, ok := p.podSyncStatuses[uid]; ok {
-		return status.IsEvicted() || (status.IsDeleted() && status.IsTerminated())
+		return status.IsEvicted() && status.IsFinished() || (status.IsDeleted() && status.IsTerminated())
```
I'm not certain that this is the right fix, but let me describe why so that others can verify or disagree.  This is a high impact line - changing it has wide ranging impacts to pod shut down.

In general, we want to remove pod content when it is no longer needed - as early as possible.  However, there are many different types of pod content - some of which are necessary for final pod status to be recorded.  We shouldn't delay *all* pod content from being removed later, just because one particular bit of content is missing.  That's why changing this line here is probably premature / incorrect. 

What we want to do is ensure that the pod content that results in status being lost is captured by the kubelet (in memory) before we remove it.   If the trigger here was eviction, it's the eviction manager's responsibility to ensure that data shows up, unless doing so would risk the kubelet's ability to remain healthy.

A bigger problem might be that we are incorrectly using the eviction manager to enforce a soft quota using a hard mechanism.  If the only pod the quota exceeded impacts is the one using the local volume, then the pod should have a chance to shut down gracefully.  That was brought up in sig-node.

I think the bug is requires a different fix than changing this line.  I think that *before* eviction cleans a resource it should capture the current status and pass that along to the pod worker (if it already has it in memory) so that the pod worker can then pass it along to the status and avoid ContainerStatusUnknown.  That's what the lines in ReconcilePod should be doing, but perhaps there's a bug before that - perhaps eviction manager needs to be able to capture the state of the container before it deletes it, or releases the content.

smarterclayton replied:
I don't think we should be changing this line of code.  Also, if changing this line of code didn't fail an e2e or unit test, we should probably fix that, because this is a pretty important line of code that we don't want to tweak without a really deep discussion.  :)

AnishShah replied:
Let me first give more context about the root cause.

1. After eviction, the kubelet kills the pod here in `SyncTerminatingPod` func. https://github.com/kubernetes/kubernetes/blob/d9c54f69d4bb7ae1bb655e1a2a50297d615025b5/pkg/kubelet/kubelet.go#L2037

2. After the pod is killed, the kubelet fetches pod status from the container runtime. https://github.com/kubernetes/kubernetes/blob/d9c54f69d4bb7ae1bb655e1a2a50297d615025b5/pkg/kubelet/kubelet.go#L2055

3. But there's a race condition because GenericPLEG sends a `ContainerDied` event upon which we clean-up the container. 

https://github.com/kubernetes/kubernetes/blob/d9c54f69d4bb7ae1bb655e1a2a50297d615025b5/pkg/kubelet/kubelet.go#L2463-L2466

Due to this race condition, the container runtime is unable to return the pod status at step 2.


IMO there are two ways we can fix this:

1. It is what I have done in this PR, that is, syncing container removal only after ensuring that it is finished. But I had a hunch that this might not be the correct fix and your comment confirms it.

2. The second way is to modify the `killPod` func to return the terminal status of the stopped pod instead of fetching it from the container runtime. There's also a TODO from you here :)

https://github.com/kubernetes/kubernetes/blob/d9c54f69d4bb7ae1bb655e1a2a50297d615025b5/pkg/kubelet/kubelet.go#L2050-L2055

hshiina replied:
I wonder if we could call `kl.podCache.Get()` and `kl.statusManager.SetPodStatus()` somehow here when a pod is being evicted:
https://github.com/kubernetes/kubernetes/blob/7c48c2bd72b9bf5c44d21d7338cc7bea77d0ad2a/pkg/kubelet/kubelet.go#L2463-L2467
At this point, a container exit code is cached by PLEG and the container is not removed yet. The exit code can be gotten from the cache and passed to the state manager.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2716,8 +2716,8 @@ func (kl *Kubelet) HandlePodReconcile(pods []*v1.Pod) {
 		if eviction.PodIsEvicted(pod.Status) {
-			if podStatus, err := kl.podCache.Get(pod.UID); err == nil {
-				kl.containerDeletor.deleteContainersInPod("", podStatus, true)
+			if _, err := kl.podCache.Get(pod.UID); err == nil {
```
I think this might actually be wrong - the comment above was last left by me because this block is indeed questionable.  I think the right action is to remove this entire if check, and find out whether evicted pods are correctly terminated with the same latency.

We should in this PR:

1. Remove this if block completely
2. Verify that with this if block removed, an evicted pod has no unnecessary delay to removal
3. If it does have delay, find out why and deliver a more targeted fix

The end outcome is that *no one* except eviction and pod worker should be deleting containers.  If we can completely remove this block safely, I believe your problem is fixed.  However, we need to do a bit more testing to ensure that is the right outcome.

AnishShah commented:
/remove-lifecycle stale

gnuletik commented:
/remove-lifecycle rotten
